source/opt/set_spec_constant_default_value_pass.cpp

```diff
@@ -248,5 +250,85 @@ bool SetSpecConstantDefaultValuePass::Process(ir::Module* module) {
   return modified;
 }
 
+// A helper function for parsing spec constant default value string. Returns
+// true if the char pointed by the given pointer is considered as the end of
+// the string, which means it is a '\0' or '}'.
+inline bool IsEnd(const char* const ptr) {
+  if (!ptr) return false;
+  return *ptr == '\0' || *ptr == '}';
+}
+
+// A helper function for parsing spec constant default value string. Returns
+// true if the char pointed by the given pointer is considered as the content
+// of an entry, which means it is not a space, end, ':' or ','.
+inline bool IsEntryContent(const char* const ptr) {
+  if (!ptr) return false;
+  return std::isspace(*ptr) == 0 && !IsEnd(ptr) && *ptr != ',' && *ptr != ':';
+}
+
+std::unique_ptr<SetSpecConstantDefaultValuePass::SpecIdToValueStrMap>
+SetSpecConstantDefaultValuePass::ParseDefaultValuesString(const char* str) {
+  if (!str) return nullptr;
+  if (*str++ != '{') return nullptr;
+
+  auto spec_id_to_value = MakeUnique<SpecIdToValueStrMap>();
+
+  // Parsing loop, break when points to the end. Do not advance pointer str in
+  // the loop without checking for '\0'.
+  while (!IsEnd(str)) {
+    // Find the spec id.
+    while (std::isspace(*str) != 0) str++;  // skip heading spaces.
+    const char* entry_begin = str;
+    while (IsEntryContent(str)) str++;
+    const char* entry_end = str;
+    std::string spec_id_str(entry_begin, entry_end - entry_begin);
+    uint32_t spec_id = 0;
+    if (!ParseNumber(spec_id_str.c_str(), &spec_id)) {
+      // The spec id is not a valid uint32 number.
+      return nullptr;
+    }
+    auto iter = spec_id_to_value->find(spec_id);
+    if (iter != spec_id_to_value->end()) {
+      // defining spec id multiple times
+      return nullptr;
+    }
+    // Find the ':'
+    while (std::isspace(*str) != 0) str++;
+    if (*str++ != ':') {
+      // ':' not found
+      return nullptr;
+    }
+    // Find the value string
+    while (std::isspace(*str) != 0) str++;
+    const char* val_begin = str;
+    while (IsEntryContent(str)) str++;
+    const char* val_end = str;
+    if (val_end == val_begin) {
+      // value string is empty.
+      return nullptr;
+    }
+    // Update the mapping with spec id and value string.
+    (*spec_id_to_value)[spec_id] = std::string(val_begin, val_end - val_begin);
+
+    // skip spaces after value string but before ',' or '}'
+    while (std::isspace(*str) != 0) str++;
+
+    // Entries must be separated by ',', if the next non-space char is not
```

Comment:
This complexity is caused by the complexity of the syntax you're accepting.  Not needed.